protocols/imap/src/main/java/org/apache/james/imap/processor/ListProcessor.java

```diff
@@ -207,30 +213,33 @@ private List<Pair<MailboxPath, ListResponse>> getListResponseForSelectSubscribed
         allSubscribedSearch.stream()
             .filter(subscribed -> !listRecursiveMatchPath.contains(subscribed))
             .filter(mailboxQuery::isPathMatch)
-            .forEach(subscribed -> {
-                ListResponse.Builder builder = ListResponse.builder()
-                    .returnSubscribed(RETURN_SUBSCRIBED);
-                if (searchedResultMap.containsKey(subscribed)) {
-                    MailboxMetaData metaData = searchedResultMap.get(subscribed);
-
-                    builder.name(mailboxName(relative, subscribed, metaData.getHierarchyDelimiter()))
-                        .children(metaData.inferiors())
-                        .selectability(metaData.getSelectability())
-                        .hierarchyDelimiter(metaData.getHierarchyDelimiter())
-                        .returnNonExistent(!RETURN_NON_EXISTENT);
-                } else {
-                    builder.name(subscribed.getName())
-                        .children(MailboxMetaData.Children.CHILDREN_ALLOWED_BUT_UNKNOWN)
-                        .selectability(MailboxMetaData.Selectability.NONE)
-                        .hierarchyDelimiter(mailboxSession.getPathDelimiter())
-                        .returnNonExistent(RETURN_NON_EXISTENT);
-                }
-                responseBuilders.add(Pair.of(subscribed, builder.build()));
-            });
+            .map(subscribed -> buildListResponse(searchedResultMap, mailboxSession, relative, subscribed))
+            .forEach(responseBuilders::add);
 
         return responseBuilders.build();
     }
 
+    private Pair<MailboxPath, ListResponse> buildListResponse(Map<MailboxPath, MailboxMetaData> searchedResultMap, MailboxSession mailboxSession, boolean relative, MailboxPath subscribed) {
+        ListResponse.Builder builder = ListResponse.builder()
+            .returnSubscribed(RETURN_SUBSCRIBED);
+        if (searchedResultMap.containsKey(subscribed)) {
+            MailboxMetaData metaData = searchedResultMap.get(subscribed);
+
+            builder.name(mailboxName(relative, subscribed, metaData.getHierarchyDelimiter()))
+                .children(metaData.inferiors())
+                .selectability(metaData.getSelectability())
+                .hierarchyDelimiter(metaData.getHierarchyDelimiter())
+                .returnNonExistent(!RETURN_NON_EXISTENT);
+        } else {
+            builder.name(subscribed.getName())
+                .children(MailboxMetaData.Children.CHILDREN_ALLOWED_BUT_UNKNOWN)
+                .selectability(MailboxMetaData.Selectability.NONE)
+                .hierarchyDelimiter(mailboxSession.getPathDelimiter())
+                .returnNonExistent(RETURN_NON_EXISTENT);
+        }
+        return Pair.of(subscribed, builder.build());
```

Comment:
We could easily do this with an Optional.

Also duplicating the first line and the last line would allow to avoid mutability.

```
return Pair.of(subscribed, 
    Optional.ofNullable(searchedResultMap.get(subscribed))
        .map(metaData -> ListResponse.builder()
                .returnSubscribed(RETURN_SUBSCRIBED)
                .name(mailboxName(relative, subscribed, metaData.getHierarchyDelimiter()))
                .children(metaData.inferiors())
                .selectability(metaData.getSelectability())
                .hierarchyDelimiter(metaData.getHierarchyDelimiter())
                .returnNonExistent(!RETURN_NON_EXISTENT))
        .orElseGet(() -> ListResponse.builder()
                .returnSubscribed(RETURN_SUBSCRIBED)
                .name(subscribed.getName())
                .children(MailboxMetaData.Children.CHILDREN_ALLOWED_BUT_UNKNOWN)
                .selectability(MailboxMetaData.Selectability.NONE)
                .hierarchyDelimiter(mailboxSession.getPathDelimiter())
                .returnNonExistent(RETURN_NON_EXISTENT))
       .build());
```

In my bucket list, from there, I would try to look how factory methods looks like:

```
ListResponse.builder()
    .nonExitingSubscribedMailbox(subscribed)
```

For instance.

Or...

```
ListResponse.builder()
    .forPath(subscribed)
    .forMetadata(metaData)
    .returnSubscribed(RETURN_SUBSCRIBED)
```

My feeling is also that by putting a bit more convenience methods in the ListResponse.Builder we could kill a lot of boiler-plate and code duplication, and make this more readable.